README.md
![Breadboard diagram](doc/fritzing.png)



## Verify/detect the presence of a device on the I2C bus

```
pi@raspberrypi:~ $ i2cdetect -y 1
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
00:                         -- -- -- -- -- -- -- --
10: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
20: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
30: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
40: 40 -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
50: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
60: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
70: 70 -- -- -- -- -- -- --
```

## Cross-compile the code for ARM architecture
NOTE: This step is automated in the VSCode task *cross: build (aarch64)*

```
user@host:~ $ export CROSS_CONTAINER_IN_CONTAINER=1
user@host:~ $ cross build --release \
                          --target aarch64-unknown-linux-gnu
```

## Copy files to rpi
NOTE: This step is automated in the VSCode task *Deploy to rpi*

```
user@host:~ $ scp target/aarch64-unknown-linux-gnu/release/pca9685-* \
                  data/* \
                  rpi:/var/tmp/
```

## Execute a channel test
```
pi@raspberrypi:~ $ export RUST_LOG=debug

# right-most angle for DS3218 20Kg servo on Channel 0
pi@raspberrypi:~ $ /var/tmp/pca9685-channel-tester --config-file-path /var/tmp/pca9685.yaml \
                                     0 \
                                     0.5

# center angle for DS3218 20Kg servo on Channel 0
pi@raspberrypi:~ $ /var/tmp/pca9685-channel-tester --config-file-path /var/tmp/pca9685.yaml \
                                     0 \
                                     1.5

# left-most angle for DS3218 20Kg servo on Channel 0
pi@raspberrypi:~ $ /var/tmp/pca9685-channel-tester --config-file-path /var/tmp/pca9685.yaml \
                                     0 \
                                     2.5
```
## Test the server
```
pi@raspberrypi:~ $ export RUST_LOG=debug
pi@raspberrypi:~ $ export ROCKET_CONFIG=/var/tmp/rocket.toml
pi@raspberrypi:~ $ /var/tmp/pca9685-service --config-file-path /var/tmp/pca9685.yaml

# In another shell...
user@host:~ $ curl http://raspberrypi.local:9999/status
{"status":"HEALTHY","software":{"version":"1.1.0"}}

# NOTE: Channel may already be configured, depending on content of
#       /var/tmp/pca9685.yaml
user@host:~ $ curl -X POST \
                   -H "Content-Type: application/json" \
                   -d @data/channel_0_config.json \
                   http://raspberrypi.local:9999/channel

# right-most angle for DS3218 20Kg servo on Channel 0
user@host:~ $ curl -X PUT \
                   -H "Content-Type: application/json" \
                   -d @data/channel_0_pw_1.5ms.json \
                   http://raspberrypi.local:9999/channel/0

```

## Prepare the host (e.g., linux/amd64) to build for target architecture (linux/arm64)

```
user@host:~ $ sudo apt-get install -y qemu-user-static
.
.
.

user@host:~ $ qemu-aarch64-static --version
qemu-aarch64 version 6.2.0 (Debian 1:6.2+dfsg-2ubuntu6.6)
Copyright (c) 2003-2021 Fabrice Bellard and the QEMU Project developers

user@host:~ $ update-binfmts --version
binfmt-support 2.2.1

user@host:~ $ docker run --rm \
                         --privileged \
                         linuxkit/binfmt:ecd24b710cad869af6a4716809f62d7164286367

user@host:~ $ docker buildx ls
NAME/NODE     DRIVER/ENDPOINT STATUS  BUILDKIT PLATFORMS
default *     docker
  default     default         running 20.10.22 linux/amd64, linux/386, linux/arm64, linux/riscv64, linux/ppc64le, linux/s390x, linux/arm/v7, linux/arm/v6
```

## Build the image for the target architecture
NOTE: Requires `docker buildx` (experimental) and QEMU on the *host* (see above)

NOTE: This step is automated in the VSCode task *build docker image (arm64)*


```
user@host:~ $ docker buildx build --platform linux/arm64 \
                                  --tag ghcr.io/kerrys-learning-lab/pca9685-service:latest-arm64 \
                                  .
```

## Manually run the image on the Raspberry pi

```
pi@raspberrypi:~ $ docker run --rm \
                              -it \
                              -e RUST_LOG=debug \
                              -e ROCKET_CONFIG=/var/tmp/rocket.toml \
                              -v /var/tmp:/var/tmp \
                              -v /dev/i2c-1:/dev/i2c-1 \
                              --privileged \
                              -p 9999:9999 \
                              ghcr.io/kerrys-learning-lab/pca9685-service:latest-arm64 \
                                --config-file-path /var/tmp/pca9685.yaml
```


CHANGELOG.md
## v1.3.0 (2023-02-18)

### Feat

- Advanced configuration (#13)

## v1.2.0 (2023-02-17)

### Feat

- Add /status endpoint (remove /ping)

## v1.1.0 (2023-01-31)

### Feat

- Create RESTful service and Docker image

## v1.0.2 (2023-01-25)

### Fix

- Update Cargo.toml:repository

## v1.0.1 (2023-01-25)

### Fix

- Convert to crate [test ci]

## v1.0.0 (2023-01-25)

### Feat

- Convert to crate

### Fix

- Update README for DS3218 servo
- Initial development


src/lib.rs
use crate::utils::{deserialize_channel, serialize_channel};
use linux_embedded_hal::i2cdev::linux::LinuxI2CError;
use pwm_pca9685::Channel;
use pwm_pca9685::OutputDriver;
use serde::Deserialize;
use serde::Serialize;
use std::collections::HashMap;
use std::sync::Mutex;

mod channelproxy;
pub mod pca9685;
mod pca9685_proxy;
pub mod utils;

/// The PCA9685 has 4096 steps/counts (12-bit PWM) of resolution
pub const PCA_PWM_RESOLUTION: u16 = 4096;

#[derive(Debug, Deserialize)]
/// An immutable YAML-based configuration of a [Pca9685] device.
pub struct Config {
    /// Path to I2C device file (e.g, /dev/i2c-1)
    pub device: String,

    /// Address of PCA9685 (e.g, 0x40)
    pub address: u8,

    /// PWM output frequency
    pub output_frequency_hz: u16,

    /// Open drain (if not set, use Totem pole)
    #[serde(default)]
    pub open_drain: bool,

    #[serde(default)]
    pub channels: Vec<ChannelConfig>,
}

#[derive(Deserialize, Serialize, PartialEq, Clone, Copy)]
/// Limits can be specified in units of counts or milliseconds (pulse-width)
///
/// Only one of count_limits or pw_limits may be supplied at configuration-time.
/// When pw_limits are given, the corresponding count_limits are automatically
/// calculated based on the configured output_frequency.
pub struct ChannelLimits {
    pub count_limits: Option<ChannelCountLimits>,
    pub pw_limits: Option<ChannelPulseWidthLimits>,
}

#[derive(Deserialize, Serialize, PartialEq, Debug, Clone, Copy)]
/// Constrains the limits of a Channel to values other than the default [0, 4095].
///
/// For example, a servo may be constrained to [1000, 3000] which then affects
/// the behavior of subsequent calls to [Pca9685::set_pwm_count],
/// [Pca9685::set_pw_ms], and [Pca9685::set_pct]
pub struct ChannelCountLimits {
    pub min_on_count: u16,
    pub max_on_count: u16,
}

#[derive(Deserialize, Serialize, PartialEq, Debug, Clone, Copy)]
/// Constrains the limits of a Channel to values other than the values
/// automatically derived from output_frequency.
///
/// For example, a servo may be constrained to [1.0ms, 2.0ms] which then affects
/// the behavior of subsequent calls to [Pca9685::set_pwm_count],
/// [Pca9685::set_pw_ms], and [Pca9685::set_pct]
pub struct ChannelPulseWidthLimits {
    pub min_on_ms: f64,
    pub max_on_ms: f64,
}

#[derive(Deserialize, Serialize, Debug)]
/// Represents the desired and/or actual configuration of a Channel.
///
/// As an input, sets the `ChannelCountLimits` on the associated Channel (in
/// which case `current_count` is not used).
///
/// As an output, describes the current PWM count (`current_count`) and
/// configured limits (`custom_limits`), if any.
pub struct ChannelConfig {
    #[serde(
        serialize_with = "serialize_channel",
        deserialize_with = "deserialize_channel"
    )]
    pub channel: Channel,
    pub current_count: Option<u16>,
    pub custom_limits: Option<ChannelLimits>,
}

#[derive(PartialEq, Debug, Clone, Copy)]
struct PcaClockConfig {
    max_pw_ms: f64,
    single_pw_duration_ms: f64,
}

struct ChannelProxy {
    name: String,
    config: ChannelConfig,
    clock_config: PcaClockConfig,
}

trait Pca9685Proxy {
    fn max_pw_ms(&self) -> f64;

    fn single_count_duration_ms(&self) -> f64;

    fn output_frequency_hz(&self) -> u16;

    fn device(&self) -> String;

    fn address(&self) -> u8;

    fn prescale(&self) -> u8;

    fn output_type(&self) -> OutputDriver;

    fn set_channel_off_count(
        &mut self,
        channel: Channel,
        off: u16,
    ) -> Result<(), pwm_pca9685::Error<LinuxI2CError>>;

    fn set_channel_full_on(
        &mut self,
        channel: Channel,
    ) -> Result<(), pwm_pca9685::Error<LinuxI2CError>>;

    fn set_channel_full_off(
        &mut self,
        channel: Channel,
    ) -> Result<(), pwm_pca9685::Error<LinuxI2CError>>;
}

/// Provides access to a PCA9685 controller, with the ability to customize the
/// range of each Channel, and set each Channel's value using raw counts,
/// pulse width in milliseconds, or percent of max pulse width.
pub struct Pca9685 {
    inner: Mutex<Box<dyn Pca9685Proxy>>,
    channels: Mutex<HashMap<u8, ChannelProxy>>,
}

/// Represents the possible errors that may occur when commanding the [Pca9685].
pub enum Pca9685Error {
    NoSuchChannelError(u8),
    PulseWidthRangeError(f64, f64),
    CustomLimitsError(u16, ChannelLimits),
    InvalidConfiguration(String),
    PercentOfRangeError(f64),
    Pca9685DriverError(pwm_pca9685::Error<LinuxI2CError>),
}

/// Customized [Result], where the error type is [Pca9685Error]
pub type Pca9685Result<T> = Result<T, Pca9685Error>;


src/main.rs
use clap::Parser;
use env_logger;
use pca9685::{Config, Pca9685};
use pwm_pca9685::Channel;
use std::fs;

/// Simple program to interact with a PCA9685
#[derive(Parser, Debug)]
#[command(author, version, about, long_about = None)]
struct Args {
    /// Channel
    #[arg(value_parser = clap::value_parser!(u8).range(..16))]
    channel: u8,

    /// Pulse width (ms)
    #[arg()]
    pulse_width_ms: f64,

    /// Path to configuration file
    #[arg(long, default_value = "/etc/pca9685.yaml")]
    config_file_path: String,
}

fn main() {
    env_logger::init();

    let args = Args::parse();

    let config = fs::read_to_string(args.config_file_path).unwrap();
    let config: Config = serde_yaml::from_str(&config).unwrap();
    let mut pca = Pca9685::new(&config);

    let channel = Channel::try_from(args.channel).unwrap();
    pca.set_pw_ms(channel, args.pulse_width_ms).unwrap();
}


src/utils.rs
use pwm_pca9685::Channel;
use serde::de::{self, Visitor};
use serde::{Deserializer, Serializer};
use std::{fmt, fs};

use crate::{
    ChannelConfig, ChannelCountLimits, ChannelLimits, ChannelPulseWidthLimits, Config,
    Pca9685Error, Pca9685Result, PcaClockConfig, PCA_PWM_RESOLUTION,
};

impl Config {
    pub fn load_from_file(path: &String) -> Config {
        let config = fs::read_to_string(path).unwrap();

        serde_yaml::from_str(&config).unwrap()
    }
}

impl ChannelConfig {
    pub fn limits(&self) -> (u16, u16) {
        match self.custom_limits {
            Some(limits) => limits.count_limits(),
            None => (0, PCA_PWM_RESOLUTION),
        }
    }
}

impl PcaClockConfig {
    pub fn pw_to_count(&self, pw_ms: f64) -> Result<u16, Pca9685Error> {
        if pw_ms < 0.0 || pw_ms > self.max_pw_ms {
            return Err(Pca9685Error::PulseWidthRangeError(pw_ms, self.max_pw_ms));
        }

        Ok((pw_ms / self.single_pw_duration_ms) as u16)
    }
}

impl ChannelLimits {
    pub fn from_count_limits(min_on_count: u16, max_on_count: u16) -> Self {
        Self {
            count_limits: Some(ChannelCountLimits {
                min_on_count: min_on_count,
                max_on_count: max_on_count,
            }),
            pw_limits: None,
        }
    }

    pub(crate) fn from_pw_limits(
        min_on_pw_ms: f64,
        max_on_pw_ms: f64,
        clock_config: PcaClockConfig,
    ) -> Self {
        Self {
            count_limits: Some(ChannelCountLimits {
                min_on_count: clock_config.pw_to_count(min_on_pw_ms).unwrap(),
                max_on_count: clock_config.pw_to_count(max_on_pw_ms).unwrap(),
            }),
            pw_limits: Some(ChannelPulseWidthLimits {
                min_on_ms: min_on_pw_ms,
                max_on_ms: max_on_pw_ms,
            }),
        }
    }

    /// Returns true if `value` is within [`min_on_count`, `max_on_count`]
    pub fn is_valid(&self, value: u16) -> bool {
        self.count_limits.unwrap().is_valid(value)
    }

    pub fn count_limits(&self) -> (u16, u16) {
        // count_limits should always be valid, because pw_limits are converted
        // to count_limits
        (
            self.count_limits.unwrap().min_on_count,
            self.count_limits.unwrap().max_on_count,
        )
    }

    pub fn pct_to_count(&self, pct: f64) -> Pca9685Result<u16> {
        if pct < 0.0 || pct > 1.0 {
            return Err(Pca9685Error::PercentOfRangeError(pct));
        }

        let (min_on_count, max_on_count) = self.count_limits();
        let pwm_range_width = max_on_count - min_on_count;
        let scaled_pwm_pct = pwm_range_width as f64 * pct;

        Ok(scaled_pwm_pct as u16 + min_on_count)
    }
}

impl fmt::Debug for ChannelLimits {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let (min_on_count, max_on_count) = self.count_limits();

        match &self.pw_limits {
            Some(pw_limits) => {
                write!(
                    f,
                    "[{}ms, {}ms) ( [{}, {}) )",
                    pw_limits.min_on_ms, pw_limits.max_on_ms, min_on_count, max_on_count
                )
            }
            None => match self.count_limits {
                Some(_) => write!(f, "[{}, {})", min_on_count, max_on_count),
                None => Ok(()),
            },
        }
    }
}

impl Default for ChannelCountLimits {
    fn default() -> Self {
        Self {
            min_on_count: 0,
            max_on_count: PCA_PWM_RESOLUTION,
        }
    }
}

impl ChannelCountLimits {
    /// Returns true if `value` is within [`min_on_count`, `max_on_count`]
    pub fn is_valid(&self, value: u16) -> bool {
        value >= self.min_on_count && value <= self.max_on_count
    }
}

impl Default for ChannelLimits {
    fn default() -> Self {
        Self {
            count_limits: Some(Default::default()),
            pw_limits: None,
        }
    }
}

impl fmt::Debug for Pca9685Error {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match &*self {
            Pca9685Error::NoSuchChannelError(channel) => write!(
                f,
                "Invalid channel: {}.  Valid channels are [0,16).",
                channel
            ),
            Pca9685Error::PulseWidthRangeError(value, max_pw_ms) => write!(
                f,
                "Pulse width value ({}ms) must be within the limits [0, {}].",
                value, max_pw_ms
            ),
            Pca9685Error::CustomLimitsError(value, limits) => write!(
                f,
                "Value ({}) must be within the limits [{}, {}].",
                value,
                limits.count_limits().0,
                limits.count_limits().1
            ),
            Pca9685Error::InvalidConfiguration(msg) => write!(f, "Invalid configuration: {}", msg),
            Pca9685Error::PercentOfRangeError(value) => write!(
                f,
                "Percentage value ({:0.4}) must be within the limits [0.0, 1.0]",
                value
            ),
            Pca9685Error::Pca9685DriverError(error) => {
                write!(
                    f,
                    "An error occurred with the underlying PCA9685 driver: {:?}",
                    error
                )
            }
        }
    }
}

impl fmt::Display for Pca9685Error {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{:?}", self)
    }
}

pub fn serialize_channel<S>(channel: &Channel, serializer: S) -> Result<S::Ok, S::Error>
where
    S: Serializer,
{
    serializer.serialize_u8(*channel as u8)
}

struct ChannelVisitor;
impl<'de> Visitor<'de> for ChannelVisitor {
    type Value = Channel;

    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
        formatter.write_str("an integer between 0 and 15, inclusive")
    }

    fn visit_u8<E>(self, value: u8) -> Result<Self::Value, E>
    where
        E: de::Error,
    {
        Ok(Channel::try_from(value).unwrap())
    }

    fn visit_u16<E>(self, value: u16) -> Result<Self::Value, E>
    where
        E: de::Error,
    {
        self.visit_u8(value as u8)
    }

    fn visit_u32<E>(self, value: u32) -> Result<Self::Value, E>
    where
        E: de::Error,
    {
        self.visit_u8(value as u8)
    }

    fn visit_u64<E>(self, value: u64) -> Result<Self::Value, E>
    where
        E: de::Error,
    {
        self.visit_u8(value as u8)
    }
}

pub fn deserialize_channel<'de, D>(deserializer: D) -> Result<Channel, D::Error>
where
    D: Deserializer<'de>,
{
    deserializer.deserialize_u8(ChannelVisitor)
}

pub mod built_info {
    // The file has been placed there by the build script.
    include!(concat!(env!("OUT_DIR"), "/built.rs"));
}


src/channelproxy.rs
use log;
use pwm_pca9685::Channel;

use crate::{
    ChannelConfig, ChannelLimits, ChannelProxy, Pca9685Error, Pca9685Proxy, Pca9685Result,
    PcaClockConfig, PCA_PWM_RESOLUTION,
};

impl ChannelProxy {
    pub fn new(channel: Channel, clock_config: PcaClockConfig) -> ChannelProxy {
        ChannelProxy {
            name: String::from(format!("Channel {:?}", channel)),
            config: ChannelConfig {
                channel: channel,
                current_count: None,
                custom_limits: None,
            },
            clock_config: clock_config,
        }
    }

    pub fn configure(&mut self, config: &ChannelConfig) -> Pca9685Result<ChannelConfig> {
        self.configure_limits(&config.custom_limits)
    }

    pub fn config(&self) -> ChannelConfig {
        ChannelConfig {
            channel: self.config.channel,
            current_count: match self.config.current_count {
                Some(pwm_count) => Some(pwm_count),
                None => None,
            },
            custom_limits: match &self.config.custom_limits {
                Some(limits) => Some(limits.clone()),
                None => None,
            },
        }
    }

    pub fn configure_limits(
        &mut self,
        custom_limits: &Option<ChannelLimits>,
    ) -> Pca9685Result<ChannelConfig> {
        match custom_limits {
            Some(limits) => {
                if limits.count_limits.is_none() && limits.pw_limits.is_none() {
                    return Err(Pca9685Error::InvalidConfiguration(
                        "ChannelConfig.custom_limits must contain either count_limits or pw_limits"
                            .to_string(),
                    ));
                }
                if limits.count_limits.is_some() && limits.pw_limits.is_some() {
                    return Err(Pca9685Error::InvalidConfiguration(
                        "ChannelConfig.custom_limits must contain only one of count_limits or pw_limits"
                            .to_string(),
                    ));
                }

                limits.count_limits.map(|count_limits| {
                    self.config.custom_limits = Some(ChannelLimits::from_count_limits(
                        count_limits.min_on_count,
                        count_limits.max_on_count,
                    ));
                });
                limits.pw_limits.map(|pw_limits| {
                    self.config.custom_limits = Some(ChannelLimits::from_pw_limits(
                        pw_limits.min_on_ms,
                        pw_limits.max_on_ms,
                        self.clock_config,
                    ));
                });

                log::info!(
                    target: &self.name,
                    "Configured limits to {:?}", self.config.custom_limits.unwrap()
                );

                Ok(self.config())
            }
            None => {
                log::info!(target: &self.name, "Configured limits to None");
                self.config.custom_limits = None;

                Ok(self.config())
            }
        }
    }

    pub fn full_on(&mut self, pca: &mut Box<dyn Pca9685Proxy>) -> Pca9685Result<ChannelConfig> {
        self.config.current_count = Some(PCA_PWM_RESOLUTION);

        log::info!(target: &self.name, "Setting output to FULL ON");

        match pca.set_channel_full_on(self.config.channel) {
            Ok(()) => Ok(self.config()),
            Err(error) => Err(Pca9685Error::Pca9685DriverError(error)),
        }
    }

    pub fn full_off(&mut self, pca: &mut Box<dyn Pca9685Proxy>) -> Pca9685Result<ChannelConfig> {
        self.config.current_count = None;

        log::info!(target: &self.name, "Setting output to FULL OFF");

        match pca.set_channel_full_off(self.config.channel) {
            Ok(()) => Ok(self.config()),
            Err(error) => Err(Pca9685Error::Pca9685DriverError(error)),
        }
    }

    pub fn set_pw_ms(
        &mut self,
        pw_ms: f64,
        pca: &mut Box<dyn Pca9685Proxy>,
    ) -> Pca9685Result<ChannelConfig> {
        self.set_pwm_count(self.clock_config.pw_to_count(pw_ms)?, pca)
    }

    pub fn set_pct(
        &mut self,
        pct: f64,
        pca: &mut Box<dyn Pca9685Proxy>,
    ) -> Pca9685Result<ChannelConfig> {
        let limits = self.config.custom_limits.unwrap_or_default();

        limits
            .pct_to_count(pct)
            .and_then(|pwm_off_count| self.set_pwm_count(pwm_off_count, pca))
    }

    pub fn set_pwm_count(
        &mut self,
        pwm_off_count: u16,
        pca: &mut Box<dyn Pca9685Proxy>,
    ) -> Pca9685Result<ChannelConfig> {
        let limits = match self.config.custom_limits {
            Some(limits) => limits,
            None => Default::default(),
        };
        if !limits.is_valid(pwm_off_count) {
            return Err(Pca9685Error::CustomLimitsError(
                pwm_off_count,
                limits.clone(),
            ));
        }

        if pwm_off_count == PCA_PWM_RESOLUTION {
            self.full_on(pca)
        } else {
            match pca.set_channel_off_count(self.config.channel, pwm_off_count) {
                Ok(()) => {
                    self.config.current_count = Some(pwm_off_count);

                    log::info!(
                        target: &self.name,
                        "Setting output to {} counts ({:0.6}ms)",
                        pwm_off_count,
                        pwm_off_count as f64 * pca.single_count_duration_ms()
                    );
                    Ok(self.config())
                }
                Err(error) => Err(Pca9685Error::Pca9685DriverError(error)),
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::{
        ChannelLimits, ChannelProxy, Pca9685Error, Pca9685Proxy, PcaClockConfig, PCA_PWM_RESOLUTION,
    };
    use pwm_pca9685::{Channel, OutputDriver};

    const TEST_OUTPUT_FREQUENCY_HZ: f64 = 200.0;
    const TEST_PCA_MAX_PW_MS: f64 = 1000.0 / TEST_OUTPUT_FREQUENCY_HZ;
    const TEST_PCA_COUNT_DURATION_MS: f64 = TEST_PCA_MAX_PW_MS / PCA_PWM_RESOLUTION as f64;
    const TEST_PCA_CLOCK_CONFIG: PcaClockConfig = PcaClockConfig {
        single_pw_duration_ms: TEST_PCA_COUNT_DURATION_MS,
        max_pw_ms: TEST_PCA_MAX_PW_MS,
    };

    struct MockPca9685Proxy;
    impl Pca9685Proxy for MockPca9685Proxy {
        fn max_pw_ms(&self) -> f64 {
            TEST_PCA_MAX_PW_MS
        }

        fn single_count_duration_ms(&self) -> f64 {
            TEST_PCA_COUNT_DURATION_MS
        }

        fn output_frequency_hz(&self) -> u16 {
            TEST_OUTPUT_FREQUENCY_HZ as u16
        }

        fn device(&self) -> String {
            String::from("/dev/foo")
        }

        fn address(&self) -> u8 {
            0x40
        }

        fn prescale(&self) -> u8 {
            31
        }

        fn output_type(&self) -> OutputDriver {
            OutputDriver::TotemPole
        }

        fn set_channel_off_count(
            &mut self,
            _channel: Channel,
            _off: u16,
        ) -> Result<(), pwm_pca9685::Error<linux_embedded_hal::i2cdev::linux::LinuxI2CError>>
        {
            Ok(())
        }

        fn set_channel_full_on(
            &mut self,
            _channel: Channel,
        ) -> Result<(), pwm_pca9685::Error<linux_embedded_hal::i2cdev::linux::LinuxI2CError>>
        {
            Ok(())
        }

        fn set_channel_full_off(
            &mut self,
            _channel: Channel,
        ) -> Result<(), pwm_pca9685::Error<linux_embedded_hal::i2cdev::linux::LinuxI2CError>>
        {
            Ok(())
        }
    }

    #[test]
    fn set_pwm_count() -> Result<(), Pca9685Error> {
        let mut channel =
            ChannelProxy::new(Channel::try_from(0 as u8).unwrap(), TEST_PCA_CLOCK_CONFIG);

        let mut mock_pca9685_proxy: Box<dyn Pca9685Proxy> = Box::new(MockPca9685Proxy {});

        // Test at min/max of range
        assert_eq!(
            channel
                .set_pwm_count(50, &mut mock_pca9685_proxy)?
                .current_count
                .unwrap(),
            50
        );

        Ok(())
    }

    #[test]
    #[should_panic(expected = "must be within the limits")]
    fn set_pwm_count_too_large() {
        let mut channel =
            ChannelProxy::new(Channel::try_from(0 as u8).unwrap(), TEST_PCA_CLOCK_CONFIG);

        let mut mock_pca9685_proxy: Box<dyn Pca9685Proxy> = Box::new(MockPca9685Proxy {});

        channel
            .set_pwm_count(PCA_PWM_RESOLUTION + 1, &mut mock_pca9685_proxy)
            .unwrap();
    }

    #[test]
    fn set_pw_ms() -> Result<(), Pca9685Error> {
        let mut channel = ChannelProxy::new(
            Channel::try_from(0 as u8).unwrap(),
            PcaClockConfig {
                single_pw_duration_ms: TEST_PCA_COUNT_DURATION_MS,
                max_pw_ms: TEST_PCA_MAX_PW_MS,
            },
        );

        let mut mock_pca9685_proxy: Box<dyn Pca9685Proxy> = Box::new(MockPca9685Proxy {});

        // Test at min/max of range
        assert_eq!(
            channel
                .set_pw_ms(0.0, &mut mock_pca9685_proxy)?
                .current_count
                .unwrap(),
            0
        );
        assert_eq!(
            channel
                .set_pw_ms(TEST_PCA_MAX_PW_MS, &mut mock_pca9685_proxy)?
                .current_count
                .unwrap(),
            4096
        );

        // Test at percentages of range
        for pct in [0.25, 0.5, 0.75] {
            let test_pw_ms = TEST_PCA_MAX_PW_MS * pct;
            let expected_counts = (4096.0 * pct) as u16;
            assert_eq!(
                channel
                    .set_pw_ms(test_pw_ms, &mut mock_pca9685_proxy)?
                    .current_count
                    .unwrap(),
                expected_counts
            );
        }

        // Test a specific value, using formula
        for test_pw_ms in [1.0, 1.5, 2.0] {
            // Hz to to millis, so to speak
            let expected_count = 1000.0 / TEST_OUTPUT_FREQUENCY_HZ as f64;

            // Duration of each count, in millis
            let expected_count = expected_count / 4096.0;

            // Number of counts required for given test_pw_ms
            let expected_count = (test_pw_ms / expected_count) as u16;

            assert_eq!(
                channel
                    .set_pw_ms(test_pw_ms, &mut mock_pca9685_proxy)?
                    .current_count
                    .unwrap(),
                expected_count
            );
        }

        return Ok(());
    }

    #[test]
    fn set_pct() -> Result<(), Pca9685Error> {
        let mut channel =
            ChannelProxy::new(Channel::try_from(0 as u8).unwrap(), TEST_PCA_CLOCK_CONFIG);

        let mut mock_pca9685_proxy: Box<dyn Pca9685Proxy> = Box::new(MockPca9685Proxy {});

        // Test at percentages of range
        for pct in [0.0, 0.25, 0.5, 0.75, 1.0] {
            let expected_counts = (4096.0 * pct) as u16;
            assert_eq!(
                channel
                    .set_pct(pct, &mut mock_pca9685_proxy)?
                    .current_count
                    .unwrap(),
                expected_counts
            );
        }

        return Ok(());
    }

    #[test]
    fn set_pct_custom_limits() -> Result<(), Pca9685Error> {
        let mut channel =
            ChannelProxy::new(Channel::try_from(0 as u8).unwrap(), TEST_PCA_CLOCK_CONFIG);

        let mut mock_pca9685_proxy: Box<dyn Pca9685Proxy> = Box::new(MockPca9685Proxy {});

        channel
            .configure_limits(&Some(ChannelLimits::from_count_limits(1000, 2000)))
            .unwrap();

        // Test at percentages of range
        for pct in [0.0, 0.25, 0.5, 0.75, 1.0] {
            let expected_counts = 1000 + (1000.0 * pct) as u16;
            assert_eq!(
                channel
                    .set_pct(pct, &mut mock_pca9685_proxy)?
                    .current_count
                    .unwrap(),
                expected_counts
            );
        }

        return Ok(());
    }

    #[test]
    #[should_panic(expected = "must be within the limits")]
    fn set_pwm_count_too_small_custom_limits() {
        let mut channel =
            ChannelProxy::new(Channel::try_from(0 as u8).unwrap(), TEST_PCA_CLOCK_CONFIG);

        channel
            .configure_limits(&Some(ChannelLimits::from_count_limits(1000, 2000)))
            .unwrap();

        let mut mock_pca9685_proxy: Box<dyn Pca9685Proxy> = Box::new(MockPca9685Proxy {});

        channel.set_pwm_count(999, &mut mock_pca9685_proxy).unwrap();
    }

    #[test]
    #[should_panic(expected = "must be within the limits")]
    fn set_pwm_count_too_large_custom_limits() {
        let mut channel =
            ChannelProxy::new(Channel::try_from(0 as u8).unwrap(), TEST_PCA_CLOCK_CONFIG);

        channel
            .configure_limits(&Some(ChannelLimits::from_count_limits(1000, 2000)))
            .unwrap();

        let mut mock_pca9685_proxy: Box<dyn Pca9685Proxy> = Box::new(MockPca9685Proxy {});

        channel
            .set_pwm_count(2001, &mut mock_pca9685_proxy)
            .unwrap();
    }

    #[test]
    #[should_panic(expected = "must be within the limits")]
    fn set_pw_ms_negative() {
        let mut channel =
            ChannelProxy::new(Channel::try_from(0 as u8).unwrap(), TEST_PCA_CLOCK_CONFIG);

        let mut mock_pca9685_proxy: Box<dyn Pca9685Proxy> = Box::new(MockPca9685Proxy {});

        channel.set_pw_ms(-1.0, &mut mock_pca9685_proxy).unwrap();
    }

    #[test]
    #[should_panic(expected = "must be within the limits")]
    fn set_pw_ms_too_large() {
        let mut channel =
            ChannelProxy::new(Channel::try_from(0 as u8).unwrap(), TEST_PCA_CLOCK_CONFIG);

        let mut mock_pca9685_proxy: Box<dyn Pca9685Proxy> = Box::new(MockPca9685Proxy {});

        channel
            .set_pw_ms(TEST_PCA_MAX_PW_MS + 1.0, &mut mock_pca9685_proxy)
            .unwrap();
    }
}
